Catch and process errors that occure during annotation instantiation

7ochem opened the pull request:
When an annotation class is instantiated and values are passed to the constructor, TypeErrors and ArgumentCountErrors can be thrown. This change catches those exceptions and shows an AnnotationException containing the annotation name and context.

Resolves: #437

stof commented on lib/Doctrine/Common/Annotations/DocParser.php:
```diff
@@ -1456,4 +1458,102 @@ private function resolvePositionalValues(array $arguments, string $name): array
+    private function instantiateAnnotiation(string $originalName, string $context, string $name, ?array $arguments = null)
+    {
+        try {
+            if ($arguments === null) {
```
couldn't you use an empty array as default value and then always use `return new $name(...$arguments);` ?

stof commented on lib/Doctrine/Common/Annotations/DocParser.php:
```diff
@@ -1456,4 +1458,102 @@ private function resolvePositionalValues(array $arguments, string $name): array
+                $matches
+            );
+            if ($matchCount === 1 && $matches[1] === $name) {
+                throw AnnotationException::creationError(sprintf(
```
Can we attach the caught error as previous exception ?

stof commented on lib/Doctrine/Common/Annotations/DocParser.php:
```diff
@@ -1456,4 +1458,102 @@ private function resolvePositionalValues(array $arguments, string $name): array
+                ));
+            }
+
+            // ArgumentCountError is not in the expected form or was not caused by the call to the `$name` constructor, so re-throw.
```
Well, the exception happened during the instantiation. So I would still always provide the context of which annotation was instantiated, to allow debugging it.
Attaching the original error as previous throwable means we can support it for any error (Saying something like `Instantiating the annotation @%s declared on %s failed.`)

stof replied:
and this would remove the need to try matching PHP error messages (which are not covered by BC rules AFAIK)

7ochem replied:
Good point. While working on this, it all started to be a bit bloated. So this would indeed be a better solution.

stof reviewed (changes requested):
This should have tests

7ochem commented:
> This should have tests

Thanks Christophe for the super fast feedback and valid points 🙏🏻 I'll work on it a bit more

7ochem committed:
Catch and process errors that occure during annotation instantiation

When an annotation class is instantiated and values are passed to the
constructor, TypeErrors and ArgumentCountErrors can be thrown. This
change catches those exceptions and shows an AnnotationException
containing the annotation name and context.

7ochem commented:
I changed the code and made it catch all `\Throwable` when instantiating the annotation class. This meant that I had to update some existing tests that were testing for an `\InvalidArgumentException` thrown in the constructor of the annotation. The advantage is that these exceptions are now also provided with the context and target.

7ochem commented on lib/Doctrine/Common/Annotations/DocParser.php:
```diff
@@ -1456,4 +1457,31 @@ private function resolvePositionalValues(array $arguments, string $name): array
+    {
+        try {
+            return new $name(...$arguments);
+        } catch (Throwable $exception) {
+            throw AnnotationException::creationError(
```
Maybe... I should add here:
```php
if ($exception instanceof AnnotationException) {
    throw $exception;
}
```

Because annotations have annotations themself. And if an exception happens in those, then you get that exception, wrapped up in an AnnotationException and that will be wrapped up in an AnnotationException (and can go on...).

stof replied:
instantiating a class will not read its doctrine annotations

7ochem replied:
That's true

stof commented on tests/Doctrine/Tests/Common/Annotations/DocParserTest.php:
```diff
@@ -897,9 +909,20 @@ public function testAnnotationEnumInvalidLiteralDeclarationException(): void
+            $parser->parse($docblock);
+        } catch (AnnotationException $exc) {
+            $previous = $exc->getPrevious();
+            $this->assertNotEmpty($previous);
```
can be removed. The `assertInstanceOf` will already reject `null`

7ochem commented on phpstan.neon:
```diff
@@ -13,6 +13,7 @@ parameters:
     ignoreErrors:
         - '#Instantiated class Doctrine_Tests_Common_Annotations_Fixtures_ClassNoNamespaceNoComment not found#'
         - '#Property Doctrine\\Tests\\Common\\Annotations\\DummyClassNonAnnotationProblem::\$foo has unknown class#'
+        - '#Call to an undefined static method PHPUnit\\Framework\\TestCase::expectExceptionMessageRegExp\(\)#'
```
Here https://github.com/doctrine/annotations/blob/065fad5dd716558d954075bbc32093cec55594e2/tests/Doctrine/Tests/Common/Annotations/DocParserTest.php#L1752-L1756

Either one of those exists: `expectExceptionMessageRegExp` exists in PHPUnit 7, deprecated in 8 and removed in 9 (and thus PHPStan gives an error). `expectExceptionMessageMatches` exists in PHPUnit 8 and up

malarzm commented:
Thanks @7ochem for the improvement and sorry for the long merge time!

7ochem commented:
> Thanks @7ochem for the improvement and sorry for the long merge time!

Thanks for merging! And no problem 👍🏻
